Fix case where `h_separation` might not work in `Button`

Rindbee opened the pull request:
<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.
-->

Fix #64140.
*Bugsquad edit:* Fixes #64371

| Before | After |
| :-------: | :-----: |
|![1](https://user-images.githubusercontent.com/30386067/183910901-a320985d-b077-40c0-8b06-0bd1b96100fd.gif)|![2](https://user-images.githubusercontent.com/30386067/183910941-8aa1eb48-bec8-4ef4-b34e-ae67183f6fe1.gif) |

~~Split from #59340.~~

There are still issues with the drawing and will be fixed in another PR.

Edit:
This patch mainly solves two things:
1. The typo of `h_separation`;
2. Negative values of `h_separation` will be treated as `0` when used, to prevent the button's minimum `width` from being reduced by `h_separation`.

KoBeWi commented on scene/gui/button.cpp:
```diff
@@ -342,16 +340,21 @@ Size2 Button::get_minimum_size_for_text_and_icon(const String &p_text, Ref<Textu
 
-	if (!expand_icon && !p_icon.is_null()) {
-		minsize.height = MAX(minsize.height, p_icon->get_height());
+	if (!p_icon.is_null()) {
```
```suggestion
	if (p_icon.is_valid()) {
```

KoBeWi commented on scene/gui/button.cpp:
```diff
@@ -342,16 +340,21 @@ Size2 Button::get_minimum_size_for_text_and_icon(const String &p_text, Ref<Textu
+			minsize.width += _icon_width_cached;
 			if (!xl_text.is_empty() || !p_text.is_empty()) {
-				minsize.width += get_theme_constant(SNAME("hseparation"));
+				minsize.width += MAX(0, get_theme_constant(SNAME("h_separation")));
```
This change makes it impossible to have negative `h_separation`. I don't think it's necessary to limit it (or at least the limitation should be documented).

Rindbee replied:
A negative value of `h_separation` is meaningless, otherwise, it may make `minsize.width` smaller, become `0` or even a **negative** value (even if there is an icon or text). In some cases this can result in insufficient visible space for drawing.

KoBeWi replied:
It makes the text overlap icon. Weird, but maybe someone wants it.
Limit it is fine too, I guess. But as I said, it should be documented in the property description.

KoBeWi reviewed (approved):
Looks good.

KoBeWi commented:
You added me as co-author (GitHub does this automatically when you accept a suggestion).

Rindbee commented:
Sorry, `h_separation` should be ignored when `expand_icon` is enabled.

KoBeWi commented:
Uh why should it be ignored? It's just a space between icon and text. If you expand icon and resize to 0, it's still there and separation should take effect. The previous behavior was alright.

KoBeWi commented:
Ah seems like CheckBox etc. apply the separation twice for their own icon.
![image](https://user-images.githubusercontent.com/2223172/184125261-2345687a-40b3-4a9c-bbb2-aff9910bce57.png)
In this case ignoring separation makes sense.

But there is still no reason to ignore it in Button. You could either add a virtual method that tells whether this should be ignored or, more simple but not much used in the codebase, some `protected bool` property.

Rindbee commented:
| version | picture |
| :-------: | :-------: |
|   v4.0.alpha13.official [82811367c] | ![alpha13](https://user-images.githubusercontent.com/30386067/184286050-7dced3be-956d-4a6f-9958-de33f035a29f.png)|
| v4.0.alpha.custom_build [317ced820] | ![0](https://user-images.githubusercontent.com/30386067/184286806-fab6e4b2-d7fd-4129-b5ae-fd8e2c33a16e.png) |
| v4.0.alpha.custom_build [323d611a6] | ![1](https://user-images.githubusercontent.com/30386067/184286062-159d0e80-3fb9-4699-b864-1f20e775b457.png) |

The table above is a comparison of the minimum size between versions. In the picture, the right node is a copy of the left node with `expand_icon` enabled.

I hope I have understood correctly what the expectation is like.

Rindbee committed:
Fix case where `h_separation` might not work in `Button`

This patch mainly solves two things:
1. The typo of `h_separation`;
2. Negative values of `h_separation` will be treated as `0` when used, to prevent the button's minimum `width` from being reduced by `h_separation`.

Rindbee commented:
I seem to be getting more and more deviated from the original intention. Now I only fix typo and restrict `h_separation` from being negative.

KoBeWi commented:
OptionButton minimum size seems broken:
![godot windows tools 64_3hNJ8pA3rH](https://user-images.githubusercontent.com/2223172/184511937-d5ba285d-ddd2-43f4-8ef8-45a13bc4dba4.gif)

CheckBox center alignment makes the text go into separation area:
![godot windows tools 64_FkKmFR9U8f](https://user-images.githubusercontent.com/2223172/184511953-be4b041d-d99e-4529-8e3a-eb7182e71de6.gif)
The worst part is that it's more to the left than Left alignment. CheckButton has similar problem.

Rindbee commented:
> OptionButton minimum size seems broken.

I guess it has something to do with `fit_to_longest_item`, switching `expand_icon` doesn't update the cache.
Steps to reproduce it:
1. Make `h_separation` larger, such as `200`;
2. Let `fit_to_longest_item` be disabled and `expand_icon` be enabled;
3. Then `enable fit_to_longest_item`;
4. Enable `expand_icon again`.

Currently, changes in factors affecting `get_minimum_size` do not update the `OptionButton` cache. I think it's better to disable `fit_to_longest_item` first and then test, finally, solve the problem of `fit_to_longest_item` cache update in other PRs.

> CheckBox center alignment makes the text go into separation area. The worst part is that it's more to the left than Left alignment. CheckButton has similar problem.

This should be a drawing issue, which will be resolved in #64351, where there is a set of comparison images with alpha13.

KoBeWi reviewed (approved):
I assume https://github.com/godotengine/godot/pull/64218#issuecomment-1211818278 is addressed in #64351?

Rindbee commented:
I guess no, both patches are trying to get back to alpha13. Except that a negative `h_separation` is treated as `0`, and limits the extent to which the `text`/`icon` can be drawn. These changes between alpha13 should be intuitively found from comparing `png`s in #64351.

This issue in https://github.com/godotengine/godot/pull/64218#issuecomment-1211818278 also exists in alpha13. 

I'm a little unclear on what exactly it should look like. So had to use alpha13 as a reference. If there is something wrong in alpha13, I will fix it later. I'm a little concerned about over-modification causing other regressions.

It might be better to figure out what it should look like first.


1. internal element (`i_icon`)
  1.1 existence
2. custom elements
  2.1. custom icon (`icon`)
    2.1.1. existence
    2.1.2. `expand_icon`
    2.1.3. `icon_alignment`
  2.2. custom text (`text`)
    2.2.1. existence 
    2.2.2. `clip_text`
    2.2.3. text `alignment`
  2.3. separation between custom elements (`s1`)
3. separation between internal element and custom elements (`s0`)

The above will directly and indirectly affect the minimum size.
In alpha13, if there are only `icon` and `i_icon`, `s0` is ignored. It's a little different when if `i_icon`, `icon`, `text` all exists.

![captrue](https://user-images.githubusercontent.com/30386067/184520153-21d98043-dd46-49d8-9c9d-00d253e9abc3.png)

![1](https://user-images.githubusercontent.com/30386067/184520255-8779b041-b4eb-46a1-a5a1-ef4cb41a7512.png)


I haven't looked into the button's design criteria, the documentation's description, the actual drawing, and your suggestions are all inconsistent.

However, this patch is intended to fix typo to get it back to a slightly better state.

akien-mga commented:
Thanks!
